[quant] Support for 4-bit quantized EmbeddingBag module

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45881 [quant] Refactor qembeddingbag to remove duplicate code
* **#45865 [quant] Support for 4-bit quantized EmbeddingBag module**
* #45752 [quant] Support 4-bit embedding_bag operators using the dtype quint4x2
* #45751 [quant] Add 4-bit embedding_bag prepack/unpack support using quint4x2

Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingStatic.test_quantized_embedding_bag
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120995](https://our.internmc.facebook.com/intern/diff/D24120995)

supriyar committed:
[quant] Support for 4-bit quantized EmbeddingBag module

Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingStatic.test_quantized_embedding_bag
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support for 4-bit quantized EmbeddingBag module"


Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingStatic.test_quantized_embedding_bag
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120995](https://our.internmc.facebook.com/intern/diff/D24120995)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support for 4-bit quantized EmbeddingBag module"


Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingStatic.test_quantized_embedding_bag
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120995](https://our.internmc.facebook.com/intern/diff/D24120995)

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/embedding_ops.py:
```diff
@@ -11,31 +11,31 @@ class EmbeddingPackedParams(torch.nn.Module):
             self._packed_weight = torch.ops.quantized.embedding_bag_prepack(weight)
         else:
-            raise RuntimeError('Unsupported dtype on quantized embedding!')
+            raise RuntimeError('Unsupported dtype for quantized embedding prepack!')
```
maybe change this to an assert?

jerryzh168 commented on torch/nn/quantized/modules/embedding_ops.py:
```diff
@@ -11,31 +11,31 @@ class EmbeddingPackedParams(torch.nn.Module):
         super(EmbeddingPackedParams, self).__init__()
         self.dtype = dtype
-        if self.dtype == torch.quint8:
+        if self.dtype in [torch.quint8, torch.quint4x2]:
```
same here, assert might be better

supriyar committed:
Update on "[quant] Support for 4-bit quantized EmbeddingBag module"


Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingStatic.test_quantized_embedding_bag
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120995](https://our.internmc.facebook.com/intern/diff/D24120995)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support for 4-bit quantized EmbeddingBag module"


Summary:

Test Plan:
python test/test_quantization.py TestPostTrainingStatic.test_quantized_embedding_bag
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120995](https://our.internmc.facebook.com/intern/diff/D24120995)

[ghstack-poisoned]
